### PR TITLE
Syncup reworking

### PR DESCRIPTION
**Pull request by nikkolasg**
This PR brings the ability to follow up a chain without being a member.
It is a bit big because it gets rid of many technical debts in the process that was touched by this feature as well.
(WIP: it does not contain cli code yet)

### CODE REVIEW

File: chain/beacon/chain.go [hunk 1]
```diff
@@ -20,40 +18,55 @@ const (
...
-		safe:          safe,
-		Store:         s,
-		done:          make(chan bool, 1),
-		ticker:        ticker,
-		newPartials:   make(chan partialInfo, defaultPartialChanBuffer),
-		newBeaconCh:   make(chan *chain.Beacon, defaultNewBeaconBuffer),
-		requestSync:   make(chan likeBeacon, defaultRequestSyncBuffer),
-		lastInserted:  make(chan *chain.Beacon, 1),
-		nonSyncBeacon: make(chan *chain.Beacon, 1),
+		l:                l,
```
Comment (willscott) [inline chain/beacon/chain.go:52]: order these in the same order as the struct definition on line 24

File: chain/beacon/chain.go [hunk 2]
```diff
@@ -20,40 +18,55 @@ const (
...
+		conf:             conf,
+		client:           client,
+		CallbackStore:    cbs,
+		sync:             syncer,
+		crypto:           crypto,
+		done:             make(chan bool, 1),
+		ticker:           ticker,
+		newPartials:      make(chan partialInfo, defaultPartialChanBuffer),
+		catchedupBeacons: make(chan *chain.Beacon, 1),
```
Comment (willscott) [inline chain/beacon/chain.go:61]: prefer `catchup` or `caughtup` to `catchedup`

File: chain/beacon/chain.go [hunk 3]
```diff
@@ -65,13 +78,10 @@ func (c *chainStore) NewValidPartial(addr string, p *drand.PartialBeaconPacket)
...
 
-func (c *chainStore) NewBeacon(addr string, proto *drand.BeaconPacket) {
-	c.newBeaconCh <- protoToBeacon(proto)
-}
-
 func (c *chainStore) Stop() {
-	c.Store.Close()
+	c.CallbackStore.Close()
 	close(c.done)
+	c.CallbackStore.RemoveCallback("chainstore")
```
Comment (willscott) [inline chain/beacon/chain.go:84]: is there a reason callbacks aren't removed by `Close`? I would expect this to either happen before calling `Close` or for `Close` to do this internally. Using objects after closing them indicates something hasn't actually stopped fully when calling Close, which is worrying

File: chain/beacon/chain.go [hunk 4]
```diff
@@ -144,70 +154,47 @@ func (c *chainStore) runAggregator() {
 				Signature:   finalSig,
 			}
 			c.l.Info("aggregated_beacon", newBeacon.Round)
-			c.newBeaconCh <- newBeacon
+			if c.tryAppend(lastBeacon, newBeacon) {
+				lastBeacon = newBeacon
+				break
+			}
+			// XXX store them for lfutur usage if it's a later round than what
```
Comment (willscott) [inline chain/beacon/chain.go:161]:
```suggestion
			// XXX store them for future usage if it's a later round than what
```

File: chain/beacon/node.go [hunk 5]
```diff
@@ -391,72 +368,66 @@ func shuffleNodes(nodes []*key.Node) []*key.Node {
...
-	}
-	return out
+	c.share = share
+	c.group = newGroup
+	c.pub = newGroup.PublicKey.PubPoly()
+	// chain info is constant
+}
+
+// CryptoSafe holds the cryptographic information to generate a partial beacon
+type CryptoSafe interface {
```
Comment (willscott) [inline chain/beacon/node.go:430]: can cryptostore (interface and implementation) get moved to a separate file?

File: chain/beacon/sync.go [hunk 6]
```diff
@@ -3,133 +3,184 @@ package beacon
...
 	}
-	defer h.l.Debug("sync_reply_leave", addr)
-	if fromRound == 0 {
-		last, err := h.chain.Last()
+	fromRound := last.Round + 1
+	// shuffle through the nodes
+	for _, n := range rand.Perm(len(nodes)) {
+		node := nodes[n]
+		// we quickly pass over different nodes to catchup fast
+		cnode, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```
Comment (willscott) [inline chain/beacon/sync.go:84]: so the connections to each node will time out their contexts in 1 second?

File: chain/beacon/chain.go (hunk 2 above)
Comment (nikkolasg) [reply chain/beacon/chain.go:61]: Lol sorry I lost my english yesterday

File: chain/beacon/chain.go (hunk 3 above)
Comment (nikkolasg) [reply chain/beacon/chain.go:84]: No you're right :)

File: chain/beacon/sync.go (hunk 6 above)
Comment (nikkolasg) [reply chain/beacon/sync.go]: Removed that line - it was an experiment I had trouble with some tests - Just add cancel now.

File: chain/beacon/node.go [hunk 7]
```diff
@@ -262,10 +237,8 @@ func (h *Handler) run(startTime int64) {
 				// network conditions allow for it.
 				// XXX find a way to start the catchup as soon as the runsync is
 				// done. Not critical but leads to faster network recovery.
-				h.l.Debug("beacon_loop", "run_sync", "potential_catchup")
-				// XXX Find a way to not run sync again before another has
-				// finished - maybe merge with chain sync mechanism
-				go h.chain.RunSync(context.Background())
+				h.l.Debug("beacon_loop", "run_sync_catchup", "last_is", lastBeacon, "should_be", current.round)
+				go h.chain.RunSync(context.Background(), current.round, nil)
```
Comment (willscott) [inline chain/beacon/node.go:241]: should this either block or be a context that can be canceled? it seems very possible that we end up with multiple of these running in parallel?  (e.g `catchup` above calls `run` and then also runs this RunSync, and this is in a loop that could select multiple times potentially, no?

Comment (nikkolasg) [reply chain/beacon/node.go:241]: The syncer doesnt starta new sync if it's already doing it so this call will return quick without doing the sync if it's already in progress.

**Comment by willscott** [comment]
i think lint is still complaining about a long line?

**Pull request merged**